### PR TITLE
Add quirk for Sonoff ZBMINIR2

### PR DESCRIPTION
**Pull request by lcheng33775823**
## Proposed change
Add quirk for Sonoff ZBMINIR2

## Checklist

- [x] The changes are tested and work correctly
- [x] `pre-commit` checks pass / the code has been formatted using Black
- [ ] Tests have been added to verify that the new code works

### CODE REVIEW

**Comment by lcheng33775823** [comment]
Hi @TheJulianJES,  I've created a PR. Is there anything I need to do before merging it?

File: zhaquirks/sonoff/zbminir2.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+
+
+(
+    QuirkBuilder("SONOFF", "ZBMINIR2")
+    .replaces(CustomSonoffCluster)
+    .enum(
+        CustomSonoffCluster.AttributeDefs.ExternalTriggerMode.name,
+        SonoffExternalSwitchTriggerType,
+        CustomSonoffCluster.cluster_id,
+        fallback_name="External Trigger Mode",
```
Comment (TheJulianJES) [inline zhaquirks/sonoff/zbminir2.py]: There are some naming conventions in Home Assistant Core that we should also follow here. ([reference](https://developers.home-assistant.io/blog/2022/07/10/entity_naming/#:~:text=the%20rest%20of%20the%20words%20are%20lower%20case)): 
```suggestion
        fallback_name="External trigger mode",
```

File: zhaquirks/sonoff/zbminir2.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+    )
+    .switch(
+        CustomSonoffCluster.AttributeDefs.TurboMode.name,
+        CustomSonoffCluster.cluster_id,
+        off_value=9,
+        on_value=20,
+        fallback_name="Turbo Mode",
```
Comment (TheJulianJES) [inline zhaquirks/sonoff/zbminir2.py]:
```suggestion
        fallback_name="Turbo mode",
```

File: zhaquirks/sonoff/zbminir2.py [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+    )
+    .switch(
+        CustomSonoffCluster.AttributeDefs.DetachRelay.name,
+        CustomSonoffCluster.cluster_id,
+        off_value=0,
+        on_value=1,
+        fallback_name="Detach Relay",
```
Comment (TheJulianJES) [inline zhaquirks/sonoff/zbminir2.py]:
```suggestion
        fallback_name="Detach relay",
```

File: zhaquirks/sonoff/zbminir2.py [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+            type=t.uint8_t,
+        )
+
+        DetachRelay = ZCLAttributeDef(
+            name="DetachRelay",
+            id=0x0017,
+            type=t.Bool,
+        )
+
+        TurboMode = ZCLAttributeDef(name="TurboMode", id=0x0012, type=t.int16s)
```
Comment (TheJulianJES) [inline zhaquirks/sonoff/zbminir2.py]: Let's remove the `name` from the `ZCLAttributeDef`. The Python attribute name is already enough.
And let's change it to be "snake case" to match other attribute definitions.

Suggestion (can be commited through GitHub UI):
```suggestion
        external_trigger_mode = ZCLAttributeDef(
            id=0x0016,
            type=t.uint8_t,
        )
        detach_relay = ZCLAttributeDef(
            id=0x0017,
            type=t.Bool,
        )
        turbo_mode = ZCLAttributeDef(
            id=0x0012,
            type=t.int16s,
        )
```

File: zhaquirks/sonoff/zbminir2.py [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+
+from zigpy import types
+from zigpy.quirks import CustomCluster
+from zigpy.quirks.v2 import QuirkBuilder
+import zigpy.types as t
+from zigpy.zcl import foundation
+from zigpy.zcl.foundation import BaseAttributeDefs, ZCLAttributeDef
+
+
+class CustomSonoffCluster(CustomCluster):
```
Comment (TheJulianJES) [inline zhaquirks/sonoff/zbminir2.py]: I think we can change this to be just `SonoffCluster`. It's a bit shorter.
```suggestion
class SonoffCluster(CustomCluster):
```

File: zhaquirks/sonoff/zbminir2.py [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+    normally_on_follow_trigger = 0x82
+
+
+(
+    QuirkBuilder("SONOFF", "ZBMINIR2")
+    .replaces(CustomSonoffCluster)
+    .enum(
+        CustomSonoffCluster.AttributeDefs.ExternalTriggerMode.name,
+        SonoffExternalSwitchTriggerType,
+        CustomSonoffCluster.cluster_id,
```
Comment (TheJulianJES) [inline zhaquirks/sonoff/zbminir2.py]: Regarding the cluster name change and attribute name change, we need to adjust the reference here too:
```suggestion
        SonoffCluster.AttributeDefs.external_trigger_mode.name,
        SonoffExternalSwitchTriggerType,
        SonoffCluster.cluster_id,
```

File: zhaquirks/sonoff/zbminir2.py [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+        fallback_name="External Trigger Mode",
+    )
+    .switch(
+        CustomSonoffCluster.AttributeDefs.TurboMode.name,
+        CustomSonoffCluster.cluster_id,
```
Comment (TheJulianJES) [inline zhaquirks/sonoff/zbminir2.py]:
```suggestion
        SonoffCluster.AttributeDefs.turbo_mode.name,
        SonoffCluster.cluster_id,
```

File: zhaquirks/sonoff/zbminir2.py [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+        off_value=9,
+        on_value=20,
+        fallback_name="Turbo Mode",
+    )
+    .switch(
+        CustomSonoffCluster.AttributeDefs.DetachRelay.name,
+        CustomSonoffCluster.cluster_id,
```
Comment (TheJulianJES) [inline zhaquirks/sonoff/zbminir2.py]:
```suggestion
        SonoffCluster.AttributeDefs.detach_relay.name,
        SonoffCluster.cluster_id,
```

File: zhaquirks/sonoff/zbminir2.py [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+    @property
+    def _is_manuf_specific(self):
+        return False
+
+
+class SonoffExternalSwitchTriggerType(types.enum8):
+    """extern switch trigger type."""
+
+    edge_trigger = 0x00
+    pluse_trigger = 0x01
```
Comment (TheJulianJES) [inline zhaquirks/sonoff/zbminir2.py]: This should probably be "pulse"?
```suggestion
    pulse_trigger = 0x01
```

File: zhaquirks/sonoff/zbminir2.py [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+    normally_off_follow_trigger = 0x02
+    normally_on_follow_trigger = 0x82
```
Comment (TheJulianJES) [inline zhaquirks/sonoff/zbminir2.py:59]: I'll have to check which naming convention we want to use here.
Other possibilities include: `EdgeTrigger`, `Edge_Trigger`, and `Edge_trigger`.

File: zhaquirks/sonoff/zbminir2.py [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+        return await super()._read_attributes(
+            attribute_ids,
+            *args,
+            manufacturer=foundation.ZCLHeader.NO_MANUFACTURER_ID,
+            **kwargs,
+        )
+
+    @property
+    def _is_manuf_specific(self):
+        return False
```
Comment (TheJulianJES) [inline zhaquirks/sonoff/zbminir2.py:49]: Hmm, in the future, we should have a better solution for this, so we don't need to duplicate it across all quirks.

**Review by TheJulianJES — COMMENTED** [review]
I've added some suggestions. I'll go ahead and commit some of them to your PR/branch.
I'll still need to check up on some other things.

File: zhaquirks/sonoff/zbminir2.py [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+
+
+class SonoffExternalSwitchTriggerType(types.enum8):
+    """extern switch trigger type."""
+
+    edge_trigger = 0x00
+    pulse_trigger = 0x01
+    normally_off_follow_trigger = 0x02
+    normally_on_follow_trigger = 0x82
```
Comment (TheJulianJES) [inline zhaquirks/sonoff/zbminir2.py]: I'll have to check which naming convention we want to use here.
Other possibilities include: `EdgeTrigger`, `Edge_Trigger`, and `Edge_trigger`.

File: zhaquirks/sonoff/zbminir2.py [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+
+
+(
+    QuirkBuilder("SONOFF", "ZBMINIR2")
+    .replaces(CustomSonoffCluster)
```
Comment (TheJulianJES) [inline zhaquirks/sonoff/zbminir2.py:63]:
```suggestion
    .replaces(SonoffCluster)
```

File: zhaquirks/sonoff/zbminir2.py [hunk 14]
```diff
@@ -0,0 +1,85 @@
...
+
+
+(
+    QuirkBuilder("SONOFF", "ZBMINIR2")
+    .replaces(SonoffCluster)
+    .enum(
+        SonoffCluster.AttributeDefs.external_trigger_mode.name,
+        SonoffExternalSwitchTriggerType,
+        SonoffCluster.cluster_id,
+        fallback_name="External trigger mode",
```
Comment (TheJulianJES) [inline zhaquirks/sonoff/zbminir2.py:69]: We now also need to add translation keys to all v2 quirk entities:
```suggestion
        translation_key="external_trigger_mode",
        fallback_name="External trigger mode",
```

File: zhaquirks/sonoff/zbminir2.py [hunk 15]
```diff
@@ -0,0 +1,85 @@
...
+    )
+    .switch(
+        SonoffCluster.AttributeDefs.turbo_mode.name,
+        SonoffCluster.cluster_id,
+        off_value=9,
+        on_value=20,
+        fallback_name="Turbo mode",
```
Comment (TheJulianJES) [inline zhaquirks/sonoff/zbminir2.py:77]:
```suggestion
        translation_key="turbo_mode",
        fallback_name="Turbo mode",
```

File: zhaquirks/sonoff/zbminir2.py [hunk 16]
```diff
@@ -0,0 +1,85 @@
...
+    )
+    .switch(
+        SonoffCluster.AttributeDefs.detach_relay.name,
+        SonoffCluster.cluster_id,
+        off_value=0,
+        on_value=1,
+        fallback_name="Detach relay",
```
Comment (TheJulianJES) [inline zhaquirks/sonoff/zbminir2.py:85]:
```suggestion
        translation_key="detach_relay",
        fallback_name="Detach relay",
```

File: zhaquirks/sonoff/zbminir2.py [hunk 17]
```diff
@@ -0,0 +1,88 @@
...
+
+
+(
+    QuirkBuilder("SONOFF", "ZBMINIR2")
+    .replaces(SonoffCluster)
+    .enum(
+        SonoffCluster.AttributeDefs.external_trigger_mode.name,
+        SonoffExternalSwitchTriggerType,
+        SonoffCluster.cluster_id,
+        translation_key="external_trigger_mode"
```
Comment (TheJulianJES) [inline zhaquirks/sonoff/zbminir2.py:68]:
```suggestion
        translation_key="external_trigger_mode",
```

File: zhaquirks/sonoff/zbminir2.py [hunk 18]
```diff
@@ -0,0 +1,88 @@
...
+        on_value=20,
+        translation_key="turbo_mode",
+        fallback_name="Turbo mode",
+    )
+    .switch(
+        SonoffCluster.AttributeDefs.detach_relay.name,
+        SonoffCluster.cluster_id,
+        off_value=0,
+        on_value=1,
+        translation_key="detach_relay"
```
Comment (TheJulianJES) [inline zhaquirks/sonoff/zbminir2.py:84]:
```suggestion
        translation_key="detach_relay",
```

**Comment by lcheng33775823** [comment]
Thank you very much for your assistance.

**Comment by pngunit** [comment]
@TheJulianJES Can this be merged? Z2M has had proper support for this device for a while, and it would be a shame to block it over variable naming.

**Comment by tomasgatial** [comment]
I have tested the Quirk, few insights:
- external trigger mode config works great 
- the detached mode config doesn't work right for me with switch in edge trigger mode: 
    switch position updates HA entity state, and HA ends up confused
- I miss inching configuration (as ewelink cluster command)

**Comment by marcohamersma** [comment]
I just got this device today, tried this quirk and can confirm it's still a bit weird when using detached mode like tomas said.

So I'm new to these types of devices, but is it normal that when using the external trigger mode of a pulse trigger, the relay switch shows up as on for a second, then reverts to false? This seems like it would make it basically impossible to distinguish between double-tapping the switch or holding the switch for a while. Wouldn't this be what the sensor is for?

File: zhaquirks/sonoff/zbminir2.py (hunk 11 above)
Comment (TheJulianJES) [reply zhaquirks/sonoff/zbminir2.py:49]: It's possible that https://github.com/zigpy/zigpy/pull/1505/ can clean this up?

File: zhaquirks/sonoff/zbminir2.py [hunk 19]
```diff
@@ -0,0 +1,88 @@
...
+        return False
+
+
+class SonoffExternalSwitchTriggerType(types.enum8):
+    """extern switch trigger type."""
+
+    edge_trigger = 0x00
+    pulse_trigger = 0x01
+    normally_off_follow_trigger = 0x02
+    normally_on_follow_trigger = 0x82
```
Comment (TheJulianJES) [inline zhaquirks/sonoff/zbminir2.py]: Renaming these to have the first letter upper-case for now. There's no real standard that we follow yet.
```suggestion
    Edge_trigger = 0x00
    Pulse_trigger = 0x01
    Normally_off_follow_trigger = 0x02
    Normally_on_follow_trigger = 0x82
```

**Review by TheJulianJES — APPROVED** [review]
The lines regarding the "no manufacturer id" modification are not covered.

I'm merging this to get initial support for the device, but we should try and use `manufacturer_id_override = foundation.ZCLHeader.NO_MANUFACTURER_ID` in the future (when https://github.com/zigpy/zigpy/pull/1505 is merged) and hopefully be able to remove those uncovered lines.
Otherwise, we should add (trivial) tests for this and/or move the `_read_attributes` and `_is_manuf_specific` overrides, as they're also required for other Sonoff devices.

Regarding the device implementation of the "detached mode", that's not affected by us. Not sure we can do anything about it with a quirk.

**Comment by arnekaiser** [comment]
I can confirm the strange behavior in detached mode. It looks like the state of the relay and also the state of the external switch are mapped to the switch entity. 

I tried the detached mode with the same device (firmware 0x00001002) with Z2M without any issues. The state of the relay is mapped to the switch entity and on external switch use, the HA device receives toggle actions.

**Comment by TheJulianJES** [comment]
Since the behavior differs from Z2M, please create an issue in this repo and link it here.

**Comment by marcohamersma** [comment]
I've just created a new issue for this in #3579

**Pull request merged**